internal/server/signup.go

```diff
@@ -0,0 +1,162 @@
+package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/gin-gonic/gin"
+
+	"github.com/infrahq/infra/api"
+	"github.com/infrahq/infra/internal"
+	"github.com/infrahq/infra/internal/access"
+	"github.com/infrahq/infra/internal/logging"
+	"github.com/infrahq/infra/internal/server/authn"
+	"github.com/infrahq/infra/internal/server/data"
+	"github.com/infrahq/infra/internal/server/email"
+	"github.com/infrahq/infra/internal/server/models"
+	"github.com/infrahq/infra/internal/server/providers"
+)
+
+func (a *API) SignupRoute() route[api.SignupRequest, *api.SignupResponse] {
+	return route[api.SignupRequest, *api.SignupResponse]{
+		handler: a.Signup,
+		routeSettings: routeSettings{
+			omitFromDocs: true,
+		},
+	}
+}
+
+func (a *API) Signup(c *gin.Context, r *api.SignupRequest) (*api.SignupResponse, error) {
+	if !a.server.options.EnableSignup {
+		return nil, fmt.Errorf("%w: signup is disabled", internal.ErrBadRequest)
+	}
+
+	keyExpires := time.Now().UTC().Add(a.server.options.SessionDuration)
+
+	var created *access.NewOrgDetails
+	switch {
+	case r.Social != nil:
+		// do social sign-up
+		if a.server.Google == nil {
+			return nil, fmt.Errorf("%w: google login is not configured, provider id must be specified for oidc login", internal.ErrBadRequest)
```

Comment:
This error isn't quite right. Probably don't need the second part

```suggestion
			return nil, fmt.Errorf("%w: google login is not configured", internal.ErrBadRequest)
```